kernel/time/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+    let func_addr = DURATION_TO_INSTANT_FUNCTION.load(Ordering::SeqCst);
+    let f: fn(Duration) -> Instant = unsafe { core::mem::transmute(func_addr) };
+    f(duration)
+}
+
+fn instant_to_duration(instant: Instant) -> Duration {
+    let func_addr = INSTANT_TO_DURATION_FUNCTION.load(Ordering::SeqCst);
+    let f: fn(Instant) -> Duration = unsafe { core::mem::transmute(func_addr) };
+    f(instant)
+}
+
+/// A measurement of a monotonically nondecreasing clock.
+///
+/// The inner value usually represents the internal counter value but the type
+/// is intentionally opaque and only useful with [`Duration`].
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct Instant {
+    counter: u64,
+}
+
+impl Instant {
+    /// Returns the amout of time elapsed from another instant to this one, or
+    /// zero duration if that instant is later than this one.
+    pub fn duration_since(&self, earlier: Self) -> Duration {
+        let instant = Instant {
+            counter: match self.counter.checked_sub(earlier.counter) {
+                Some(value) => value,
+                None => return Duration::ZERO,
+            },
+        };
+        instant_to_duration(instant)
+    }
+}
+
+impl core::ops::Add<Duration> for Instant {
+    type Output = Self;
+
+    fn add(self, rhs: Duration) -> Self::Output {
+        let instant = duration_to_instant(rhs);
+        Self {
+            counter: self.counter + instant.counter,
+        }
+    }
+}
+
+impl core::ops::Sub<Duration> for Instant {
+    type Output = Self;
+
+    fn sub(self, rhs: Duration) -> Self::Output {
+        let instant = duration_to_instant(rhs);
+        Self {
+            counter: self.counter - instant.counter,
+        }
+    }
+}
+
+impl core::ops::Sub<Instant> for Instant {
+    type Output = Duration;
+
+    fn sub(self, rhs: Instant) -> Self::Output {
+        self.duration_since(rhs)
+    }
+}
+
+/// Register the clock source that can be used to sleep when interrupts are
+/// disabled.
+///
+/// The current early sleeper will only be overwritten by `T` if `frequency` is
+/// larger than the frequency of the current early sleeper.
+///
+/// Returns whether the early sleeper was overwritten.
+///
+/// # Errors
+///
+/// Returns an error if [`INIT_REQUIRED`](EarlySleeper::INIT_REQUIRED) is `true`
+/// and [`ClockSource::init`] returned an error.
+pub fn register_early_sleeper<T>(frequency: u64) -> Result<bool, &'static str>
+where
+    T: EarlySleeper,
+{
+    let old_frequency = EARLY_SLEEPER_FREQUENCY.load(Ordering::SeqCst);
+    if frequency > old_frequency {
+        if T::INIT_REQUIRED {
+            // FIXME: The source may be double initialised: once here and once in
+            // register_clock_source. This doesn't currently cause issues but needs to be
+            // fixed. This will probably be fixed if we separate interrupts from clocks in a
+            // future PR.
+            T::init()?;
```

Comment:
hmm, isn't this backwards from what we discussed a couple weeks ago? By that I mean that a given clocksource would be initialized separately (e.g., by the device manager) and then as part of its initialization routine, it would register itself as a clocksource/early sleeper.

Personally I find this to be a bit confusing in terms of control flow.